crates/sparrow-main/src/serve/file_service.rs

```diff
@@ -55,48 +60,74 @@ impl FileService for FileServiceImpl {
 }
 
 async fn get_metadata(
-    s3: S3Helper,
+    object_store_registry: Arc<ObjectStoreRegistry>,
     request: tonic::Request<GetMetadataRequest>,
 ) -> anyhow::Result<tonic::Response<GetMetadataResponse>> {
     let request = request.into_inner();
 
     let file_metadatas: Vec<_> = request
         .file_paths
         .iter()
-        .map(|source| get_source_metadata(&s3, source))
+        .map(|source| get_source_metadata(object_store_registry.as_ref(), source))
         .collect();
 
-    let file_metadatas = try_join_all(file_metadatas).await?;
+    let file_metadatas = try_join_all(file_metadatas)
+        .await
+        .map_err(|e| anyhow::anyhow!("unable to get file metadata: {:?}", e))?;
     Ok(Response::new(GetMetadataResponse { file_metadatas }))
 }
 
+#[derive(derive_more::Display, Debug)]
+pub enum Error {
+    #[display(fmt = "unable to get source path from request")]
+    SourcePathError,
+    #[display(fmt = "unable to create temporary file")]
+    TempFileDownloadFileError,
+    #[display(fmt = "unable to download file from object store")]
+    ObjectStoreError,
+    #[display(fmt = "schema error: '{_0}'")]
+    SchemaError(String),
+}
+impl error_stack::Context for Error {}
+
 pub(crate) async fn get_source_metadata(
-    s3: &S3Helper,
+    object_store_registry: &ObjectStoreRegistry,
     source: &FilePath,
-) -> anyhow::Result<FileMetadata> {
+) -> error_stack::Result<FileMetadata, Error> {
     let source = source
         .path
         .as_ref()
-        .context(context_code!(Code::InvalidArgument, "Missing source_path"))?;
+        .context(context_code!(Code::InvalidArgument, "Missing source_path"))
```

Comment:
`context` comes from `use anyhow::Context` - let's try to remove anyhow here if we're swapping to error_stack. So can just remove this line basically. And if you want more context, can add a string to the `Error::SourcePath`